---
title: "Driver_classif_between_IDH_grade"
author: "AlizÃ©e LANON"
date: "2024-04-08"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(
  warning = FALSE,     # Suppress warnings from R code
  message = FALSE     # Default height of plots
)
```
## Grade
###### Import 
```{r, message=FALSE}
library(ggplot2)
library(BiocStyle)

#--> Phenotypic matrix
pdata <- read.csv("pdata/pdata_DA_GB_IDHwt_mut2.csv", sep =";", header = TRUE)
#pdata <- pdata[-6,]
row.names(pdata) <- pdata$Sample

#--> Count matrix
mrna_raw_counts <- read.table("../../../Glioma/Stepniak/RNAseq/nextflow/star_salmon/salmon.merged.gene_counts.tsv", header = TRUE)
row.names(mrna_raw_counts) <- mrna_raw_counts$gene_id
mrna_raw_counts <- mrna_raw_counts[,-c(1:2)]
mrna_raw_counts <- mrna_raw_counts[,c(1,6,7,2,3,4,12,13,14,15,17)]
##(1,6,7,2,3,4,12,12,14,15,16)
#--> Order and match
mrna_raw_counts <- mrna_raw_counts[,match(pdata$Sample, colnames(mrna_raw_counts), nomatch = 0)]
pdata <- pdata[pdata$Sample %in% colnames(mrna_raw_counts),]

#--> Feature annotation
mrna_gene <- read.table("../../../Glioma/Stepniak/RNAseq/nextflow/star_salmon/salmon.merged.gene_counts.tsv", header = TRUE)[,c(1:2)]
```
####### Filter ######
```{r, message=FALSE}
#--> Filter lowly expressed genes
library(HTSFilter)

mrna_filter <- HTSFilter(round(mrna_raw_counts),
                         pdata$Group,
                         s.min=1, s.max=200)
```
###### Normalization ######
```{r, message=FALSE}
library(RUVSeq)
library(reshape)
library(EDASeq)

#--> Quantile normalization
raw_set <- newSeqExpressionSet(counts = as.matrix(mrna_filter$filteredData))
colnames(raw_set) <- pdata$Sample
qn_set <- betweenLaneNormalization(raw_set, which = "full")

#--> edgeR TMM normalization
library(edgeR)

tmm_set <-  mrna_filter$filteredData

pdata_tmm <- pdata
design <- cbind(model.matrix(~ 0 + pdata_tmm$Group))
colnames(design) <- gsub(".*\\$|)", "", colnames(design))

tmm_set <- DGEList(counts = tmm_set,
                   genes = mrna_gene[match(
                     row.names(mrna_filter$filteredData), 
                     mrna_gene$gene_id),],
                   group = pdata_tmm$Group)

tmm_set <- calcNormFactors(tmm_set)

#--> Estimate dispersion and fit the model
tmm_set <- estimateGLMRobustDisp(tmm_set, design)
```
##### Quality Check
```{r, message=FALSE}
##### Boxplot
library(plotly)
#--> Get data
bp_full <- rbind(melt(log2(mrna_filter$filteredData + 1)),
                 melt(log2(normCounts(qn_set) + 1)),
                 melt(cpm(tmm_set, normalized.lib.sizes = TRUE, log = TRUE)))

#--> Build the data.frame
bp_full$Method <- rep(c("Before QN", "After QN", "After TMM"),
                      each = nrow(mrna_filter$filteredData) *    ncol(mrna_filter$filteredData))

bp_full$Method <- factor(bp_full$Method, levels = c("Before QN",
                                                    "After QN",
                                                    "After TMM"))

colnames(bp_full) <- c("Gene", "Sample", "Expression", "Method")

bp_full <- merge(bp_full, pdata,
                 by.x = "Sample", by.y = "Sample",
                 all = TRUE)

#--> Compute statistics
calc_stat <- function(x) {
  coef <- 1.5
  n <- sum(!is.na(x))
  stats <- quantile(x, probs = c(0.1, 0.25, 0.5, 0.75, 0.9))
  names(stats) <- c("ymin", "lower", "middle", "upper", "ymax")
  return(stats)
}

#--> Build the plot                 
bp <- ggplot(bp_full, aes(x = Sample, y = Expression, fill = Group, text = Sample)) + 
  stat_summary(fun.data = calc_stat, geom = "boxplot") +
  facet_wrap(~Method, scales = "free_y") +
  ggtitle("Distribution of gene expression")  +
  ylab("Expression (logCPM)") + xlab("") +
  theme() +  # Call the theme function with parentheses
  theme(axis.text.x = element_text(angle = 90, hjust = 1), legend.position = "none")


ggplotly(bp)

```
####### PCA prcomp
```{r, message=FALSE}
library(ggrepel)

#-->  Creating the expression matrices
data_pca_before_qn <- log2(mrna_filter$filteredData + 1)
data_pca_after_qn  <- log2(normCounts(qn_set) + 1)
data_pca_after_tmm <- cpm(tmm_set, normalized.lib.sizes = TRUE, log = TRUE)

#-->  Computing PCA
pca_before_qn   <- as.data.frame(prcomp(t(na.omit(data_pca_before_qn)), scale. = TRUE, center = TRUE)$x)
pca_after_qn    <- as.data.frame(prcomp(t(na.omit(data_pca_after_qn)), scale. = TRUE, center = TRUE)$x) 
pca_after_tmm   <- as.data.frame(prcomp(t(na.omit(data_pca_after_tmm)), scale. = TRUE, center = TRUE)$x)

#-->  Combine
comb <- data.frame(rbind(
  pca_before_qn[,c(1:2)],
  pca_after_qn[,c(1:2)],
  pca_after_tmm[,c(1:2)]),
  PCA = rep(c("Before QN", "After QN",
              "After TMM"), each = nrow(pdata)),
  Sample = pdata$Sample,
  Group = pdata$Group)
colnames(comb) <- c("PC1", "PC2", "PCA", "Sample","Group")

comb$PCA <- factor(comb$PCA, levels = c("Before QN", "After QN",
                                        "After TMM"))

#-->  Plot
pca_plot <- ggplot(comb, aes(x= PC1, y = PC2)) + 
  geom_point(aes(color = Group,
                 text = Sample)) +
  facet_wrap(~PCA, nrow = 2, scales = "free") +
  ggtitle("PCA - Normalization")  +
  theme()

ggplotly(pca_plot)
```
####### PCA pca
```{r, message=FALSE}
library(ggrepel)

#-->  Creating the expression matrices
data_pca_before_qn <- log2(mrna_filter$filteredData + 1)
data_pca_after_qn  <- log2(normCounts(qn_set) + 1)
data_pca_after_tmm <- cpm(tmm_set, normalized.lib.sizes = TRUE, log = TRUE)

#-->  Computing PCA
res_data_pca_before_qn <- PCA(t(na.omit(data_pca_before_qn)), scale.unit = TRUE,graph = FALSE)

plot_data_pca_before_qn <- fviz_pca_ind(res_data_pca_before_qn, title = "Pca of DA samples before qn",
             col.ind = pdata$Group, # color by groups
             palette = c("#00AFBB", "magenta"),
             legend.title = "Groups",
             ylim = c(-200, 100), xlim = c(-200,200))

res_data_pca_after_qn <- PCA(t(na.omit(data_pca_after_qn)), scale.unit = TRUE,graph = FALSE)

plot_data_pca_after_qn <- fviz_pca_ind(res_data_pca_after_qn, title = "Pca of DA samples after qn",
             col.ind = pdata$Group, # color by groups
             palette = c("#00AFBB", "magenta"),
             legend.title = "Groups",
             ylim = c(-100, 250), xlim = c(-200,150))

res_data_pca_after_tmm <- PCA(t(na.omit(data_pca_after_tmm)), scale.unit = TRUE,graph = FALSE)

plot_data_pca_after_tmm <- fviz_pca_ind(res_data_pca_after_tmm, title = "Pca of DA samples after tmm",
             col.ind = pdata$Group, # color by groups
             palette = c("#00AFBB", "magenta"),
             legend.title = "Groups",
             ylim = c(-210, 75), xlim = c(-250,150))

## Plot 
par(mfrow = c(1, 3))
print(plot_data_pca_before_qn)
print(plot_data_pca_after_qn)
print(plot_data_pca_after_tmm)

```
## IDH mutation
###### Import 
```{r, message=FALSE}
library(ggplot2)
library(BiocStyle)

#--> Phenotypic matrix
pdata <- read.csv("pdata/pdata_DA_GB_IDHwt_mut.csv", sep =";", header = TRUE)
#pdata <- pdata[-6,]
row.names(pdata) <- pdata$Sample

#--> Count matrix
mrna_raw_counts <- read.table("../../../Glioma/Stepniak/RNAseq/nextflow/star_salmon/salmon.merged.gene_counts.tsv", header = TRUE)
row.names(mrna_raw_counts) <- mrna_raw_counts$gene_id
mrna_raw_counts <- mrna_raw_counts[,-c(1:2)]
mrna_raw_counts <- mrna_raw_counts[,c(1,6,7,2,3,4,12,13,14,15,17)]
##(1,6,7,2,3,4,12,12,14,15,16)
#--> Order and match
mrna_raw_counts <- mrna_raw_counts[,match(pdata$Sample, colnames(mrna_raw_counts), nomatch = 0)]
pdata <- pdata[pdata$Sample %in% colnames(mrna_raw_counts),]

#--> Feature annotation
mrna_gene <- read.table("../../../Glioma/Stepniak/RNAseq/nextflow/star_salmon/salmon.merged.gene_counts.tsv", header = TRUE)[,c(1:2)]
```
####### Filter ######
```{r, message=FALSE}
#--> Filter lowly expressed genes
library(HTSFilter)

mrna_filter <- HTSFilter(round(mrna_raw_counts),
                         pdata$Group,
                         s.min=1, s.max=200)
```
###### Normalization ######
```{r, message=FALSE}
library(RUVSeq)
library(reshape)
library(EDASeq)

#--> Quantile normalization
raw_set <- newSeqExpressionSet(counts = as.matrix(mrna_filter$filteredData))
colnames(raw_set) <- pdata$Sample
qn_set <- betweenLaneNormalization(raw_set, which = "full")

#--> edgeR TMM normalization
library(edgeR)

tmm_set <-  mrna_filter$filteredData

pdata_tmm <- pdata
design <- cbind(model.matrix(~ 0 + pdata_tmm$Group))
colnames(design) <- gsub(".*\\$|)", "", colnames(design))

tmm_set <- DGEList(counts = tmm_set,
                   genes = mrna_gene[match(
                     row.names(mrna_filter$filteredData), 
                     mrna_gene$gene_id),],
                   group = pdata_tmm$Group)

tmm_set <- calcNormFactors(tmm_set)

#--> Estimate dispersion and fit the model
tmm_set <- estimateGLMRobustDisp(tmm_set, design)
```
##### Quality Check
```{r, message=FALSE}
##### Boxplot
library(plotly)
#--> Get data
bp_full <- rbind(melt(log2(mrna_filter$filteredData + 1)),
                 melt(log2(normCounts(qn_set) + 1)),
                 melt(cpm(tmm_set, normalized.lib.sizes = TRUE, log = TRUE)))

#--> Build the data.frame
bp_full$Method <- rep(c("Before QN", "After QN", "After TMM"),
                      each = nrow(mrna_filter$filteredData) *    ncol(mrna_filter$filteredData))

bp_full$Method <- factor(bp_full$Method, levels = c("Before QN",
                                                    "After QN",
                                                    "After TMM"))

colnames(bp_full) <- c("Gene", "Sample", "Expression", "Method")

bp_full <- merge(bp_full, pdata,
                 by.x = "Sample", by.y = "Sample",
                 all = TRUE)

#--> Compute statistics
calc_stat <- function(x) {
  coef <- 1.5
  n <- sum(!is.na(x))
  stats <- quantile(x, probs = c(0.1, 0.25, 0.5, 0.75, 0.9))
  names(stats) <- c("ymin", "lower", "middle", "upper", "ymax")
  return(stats)
}

#--> Build the plot                 
bp <- ggplot(bp_full, aes(x = Sample, y = Expression, fill = Group, text = Sample)) + 
  stat_summary(fun.data = calc_stat, geom = "boxplot") +
  facet_wrap(~Method, scales = "free_y") +
  ggtitle("Distribution of gene expression")  +
  ylab("Expression (logCPM)") + xlab("") +
  theme() +  # Call the theme function with parentheses
  theme(axis.text.x = element_text(angle = 90, hjust = 1), legend.position = "none")


ggplotly(bp)

```
####### PCA prcomp
```{r, message=FALSE}
library(ggrepel)

#-->  Creating the expression matrices
data_pca_before_qn <- log2(mrna_filter$filteredData + 1)
data_pca_after_qn  <- log2(normCounts(qn_set) + 1)
data_pca_after_tmm <- cpm(tmm_set, normalized.lib.sizes = TRUE, log = TRUE)

#-->  Computing PCA
pca_before_qn   <- as.data.frame(prcomp(t(na.omit(data_pca_before_qn)), scale. = TRUE, center = TRUE)$x)
pca_after_qn    <- as.data.frame(prcomp(t(na.omit(data_pca_after_qn)), scale. = TRUE, center = TRUE)$x) 
pca_after_tmm   <- as.data.frame(prcomp(t(na.omit(data_pca_after_tmm)), scale. = TRUE, center = TRUE)$x)

#-->  Combine
comb <- data.frame(rbind(
  pca_before_qn[,c(1:2)],
  pca_after_qn[,c(1:2)],
  pca_after_tmm[,c(1:2)]),
  PCA = rep(c("Before QN", "After QN",
              "After TMM"), each = nrow(pdata)),
  Sample = pdata$Sample,
  Group = pdata$Group)
colnames(comb) <- c("PC1", "PC2", "PCA", "Sample","Group")

comb$PCA <- factor(comb$PCA, levels = c("Before QN", "After QN",
                                        "After TMM"))

#-->  Plot
pca_plot <- ggplot(comb, aes(x= PC1, y = PC2)) + 
  geom_point(aes(color = Group,
                 text = Sample)) +
  facet_wrap(~PCA, nrow = 2, scales = "free") +
  ggtitle("PCA - Normalization")  +
  theme()

ggplotly(pca_plot)
```
####### PCA pca
```{r, message=FALSE}
library(ggrepel)

#-->  Creating the expression matrices
data_pca_before_qn <- log2(mrna_filter$filteredData + 1)
data_pca_after_qn  <- log2(normCounts(qn_set) + 1)
data_pca_after_tmm <- cpm(tmm_set, normalized.lib.sizes = TRUE, log = TRUE)

#-->  Computing PCA
res_data_pca_before_qn <- PCA(t(na.omit(data_pca_before_qn)), scale.unit = TRUE,graph = FALSE)

plot_data_pca_before_qn <- fviz_pca_ind(res_data_pca_before_qn, title = "Pca of DA samples before qn",
             col.ind = pdata$Group, # color by groups
             palette = c("#00AFBB", "magenta"),
             legend.title = "Groups",
             ylim = c(-200, 100), xlim = c(-200,200))

res_data_pca_after_qn <- PCA(t(na.omit(data_pca_after_qn)), scale.unit = TRUE,graph = FALSE)

plot_data_pca_after_qn <- fviz_pca_ind(res_data_pca_after_qn, title = "Pca of DA samples after qn",
             col.ind = pdata$Group, # color by groups
             palette = c("#00AFBB", "magenta"),
             legend.title = "Groups",
             ylim = c(-100, 250), xlim = c(-200,150))

res_data_pca_after_tmm <- PCA(t(na.omit(data_pca_after_tmm)), scale.unit = TRUE,graph = FALSE)

plot_data_pca_after_tmm <- fviz_pca_ind(res_data_pca_after_tmm, title = "Pca of DA samples after tmm",
             col.ind = pdata$Group, # color by groups
             palette = c("#00AFBB", "magenta"),
             legend.title = "Groups",
             ylim = c(-210, 75), xlim = c(-250,150))

## Plot 
par(mfrow = c(1, 3))
print(plot_data_pca_before_qn)
print(plot_data_pca_after_qn)
print(plot_data_pca_after_tmm)

```