---
title: "Differential_Analysis_Stepniak_ChiP_H3K4me3_0,1"
author: "AlizÃ©e LANON"
date: "2024-05-06"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(
  warning = FALSE,     # Suppress warnings from R code
  message = FALSE     
)
```

###### Import 
```{r, message=FALSE}
library(ggplot2)
library(BiocStyle)

#--> Phenotypic matrix 
# DA02_RNAseq2_S13;IDHwt
pdata <- read.csv("pdata_chip_H3K4me3.csv", sep =";", header = TRUE)
#pdata <- pdata[-6,]
row.names(pdata) <- pdata$Sample

#--> Count matrix
# /shared/projects/gliomep53/Scripts_R/Differential_Analysis_Stepniak/CHiP/
mrna_raw_counts <- read.delim(file = "H3K4me3.consensus_peaks.featureCounts.txt", header = TRUE, sep = "\t")
row.names(mrna_raw_counts) <- mrna_raw_counts$Geneid
mrna_raw_counts_table <- mrna_raw_counts[,c(7,8,9,10,11,12)]

#--> Order and match
mrna_raw_counts_table <- mrna_raw_counts[,match(pdata$Sample, colnames(mrna_raw_counts), nomatch = 0)]
pdata <- pdata[pdata$Sample %in% colnames(mrna_raw_counts),]

mrna_raw_counts_info <- mrna_raw_counts[,c(1,2,3,4,5)]

#--> Feature annotation
#--> Feature annotation
mrna_gene <- read.table("H3K4me3.consensus_peaks.featureCounts.txt", header = TRUE)[,c(1)]
```
####### Filter ######
```{r, message=FALSE}
#--> Filter lowly expressed genes
library(HTSFilter)

mrna_filter <- HTSFilter(round(mrna_raw_counts_table),
                         pdata$Group,
                         s.min=1, s.max=200)
```
###### Normalization ######
```{r, message=FALSE}
library(RUVSeq)
library(reshape)
library(EDASeq)

#--> Quantile normalization
raw_set <- newSeqExpressionSet(counts = as.matrix(mrna_filter$filteredData))
colnames(raw_set) <- pdata$Sample
qn_set <- betweenLaneNormalization(raw_set, which = "full")

#--> edgeR TMM normalization
library(edgeR)

tmm_set <-  mrna_filter$filteredData

pdata_tmm <- pdata
design <- cbind(model.matrix(~ 0 + pdata_tmm$Group))
colnames(design) <- gsub(".*\\$|)", "", colnames(design))

tmm_set <- DGEList(counts = tmm_set,
                   genes = mrna_gene,
                   group = pdata_tmm$Group)

tmm_set <- calcNormFactors(tmm_set)

#--> Estimate dispersion and fit the model
tmm_set <- estimateGLMRobustDisp(tmm_set, design)
```
##### Quality Check
```{r, message=FALSE}
##### Boxplot
library(plotly)
#--> Get data
bp_full <- rbind(melt(log2(mrna_filter$filteredData + 1)),
                 melt(log2(normCounts(qn_set) + 1)),
                 melt(cpm(tmm_set, normalized.lib.sizes = TRUE, log = TRUE)))

#--> Build the data.frame
bp_full$Method <- rep(c("Before QN", "After QN", "After TMM"),
                      each = nrow(mrna_filter$filteredData) *    ncol(mrna_filter$filteredData))

bp_full$Method <- factor(bp_full$Method, levels = c("Before QN",
                                                    "After QN",
                                                    "After TMM"))

colnames(bp_full) <- c("Gene", "Sample", "Expression", "Method")

bp_full <- merge(bp_full, pdata,
                 by.x = "Sample", by.y = "Sample",
                 all = TRUE)

#--> Compute statistics
calc_stat <- function(x) {
  coef <- 1.5
  n <- sum(!is.na(x))
  stats <- quantile(x, probs = c(0.1, 0.25, 0.5, 0.75, 0.9))
  names(stats) <- c("ymin", "lower", "middle", "upper", "ymax")
  return(stats)
}

#--> Build the plot                 
bp <- ggplot(bp_full, aes(x = Sample, y = Expression, fill = Group, text = Sample)) + 
  stat_summary(fun.data = calc_stat, geom = "boxplot") +
  facet_wrap(~Method, scales = "free_y") +
  ggtitle("Distribution of gene expression")  +
  ylab("Expression (logCPM)") + xlab("") +
  theme() +  # Call the theme function with parentheses
  theme(axis.text.x = element_text(angle = 90, hjust = 1), legend.position = "none")


ggplotly(bp)

```
####### PCA prcomp
```{r, message=FALSE}
library(ggrepel)
library(FactoMineR)
library(factoextra)
#-->  Creating the expression matrices
data_pca_before_qn <- log2(mrna_filter$filteredData + 1)
data_pca_after_qn  <- log2(normCounts(qn_set) + 1)
data_pca_after_tmm <- cpm(tmm_set, normalized.lib.sizes = TRUE, log = TRUE)

#-->  Computing PCA
pca_before_qn   <- as.data.frame(prcomp(t(na.omit(data_pca_before_qn)), scale. = TRUE, center = TRUE)$x)

pca_after_qn    <- as.data.frame(prcomp(t(na.omit(data_pca_after_qn)), scale. = FALSE, center = TRUE)$x) 

pca_after_tmm   <- as.data.frame(prcomp(t(na.omit(data_pca_after_tmm)), scale. = TRUE, center = TRUE)$x)

#-->  Combine
comb <- data.frame(rbind(
  pca_before_qn[,c(1:2)],
  pca_after_qn[,c(1:2)],
  pca_after_tmm[,c(1:2)]),
  PCA = rep(c("Before QN", "After QN",
              "After TMM"), each = nrow(pdata)),
  Sample = pdata$Sample,
  Group = pdata$Group)
colnames(comb) <- c("PC1", "PC2", "PCA", "Sample","Group")

comb$PCA <- factor(comb$PCA, levels = c("Before QN", "After QN",
                                        "After TMM"))

#-->  Plot
pca_plot <- ggplot(comb, aes(x= PC1, y = PC2)) + 
  geom_point(aes(color = Group,
                 text = Sample)) +
  facet_wrap(~PCA, nrow = 2, scales = "free") +
  ggtitle("PCA - Normalization")  +
  theme()

ggplotly(pca_plot)
```
####### PCA pca
```{r, message=FALSE}
library(ggrepel)
library(FactoMineR)
library(factoextra)
#-->  Creating the expression matrices
data_pca_before_qn <- log2(mrna_filter$filteredData + 1)
data_pca_after_qn  <- log2(normCounts(qn_set) + 1)
data_pca_after_tmm <- cpm(tmm_set, normalized.lib.sizes = TRUE, log = TRUE)

#-->  Computing PCA
res_data_pca_before_qn <- PCA(t(na.omit(data_pca_before_qn)), scale.unit = TRUE,graph = FALSE)

plot_data_pca_before_qn <- fviz_pca_ind(res_data_pca_before_qn, title = "Pca of DA samples before qn",
             col.ind = pdata$Group, # color by groups
             palette = c("#00AFBB", "magenta"),
             legend.title = "Groups",
             ylim = c(-200, 200), xlim = c(-350,250))

res_data_pca_after_qn <- PCA(t(na.omit(data_pca_after_qn)), scale.unit = TRUE,graph = FALSE)

plot_data_pca_after_qn <- fviz_pca_ind(res_data_pca_after_qn, title = "Pca of DA samples after qn",
             col.ind = pdata$Group, # color by groups
             palette = c("#00AFBB", "magenta"),
             legend.title = "Groups",
             ylim = c(-200, 150), xlim = c(-250,250))

res_data_pca_after_tmm <- PCA(t(na.omit(data_pca_after_tmm)), scale.unit = TRUE,graph = FALSE)

plot_data_pca_after_tmm <- fviz_pca_ind(res_data_pca_after_tmm, title = "Pca of DA samples after tmm",
             col.ind = pdata$Group, # color by groups
             palette = c("#00AFBB", "magenta"),
             legend.title = "Groups",
             ylim = c(-200, 200), xlim = c(-250,250))

## Plot 
par(mfrow = c(1, 3))
print(plot_data_pca_before_qn)
print(plot_data_pca_after_qn)
print(plot_data_pca_after_tmm)

```
###### Differential expression analysis
```{r, message=FALSE}
### ANOVA like test
#--> Estimate dispersion and fit the model
tmm_fit <- glmFit(tmm_set, design = design)

################################################
#--> ANOVA-like test
lrt_full <- glmLRT(tmm_fit,
                   contrast =  makeContrasts(IDHmut_vs_IDHwt = GroupIDHmut - GroupIDHwt,
                                             levels = design))

#--> List DE genes and correct for multiple testing
top_full <- topTags(lrt_full, n = Inf)$table
top <- topTags(lrt_full, adjust.method = "fdr",
               p.value = 0.1,  n = Inf)$table

# Extract fold change values and gene names
fold_changes <- top$logFC
gene_names <- top$genes

# Identify up-regulated and down-regulated genes
upregulated_loci <- gene_names[fold_changes > 0]
up_loci_Grange <- mrna_raw_counts_info[upregulated_loci,]

downregulated_loci <- gene_names[fold_changes < 0]
down_loci_Grange <- mrna_raw_counts_info[downregulated_loci,]
```
###### Hierarchical clustering
```{r, message=FALSE}
library(pheatmap)
library(viridis)

#--> Select ID of DE genes
sel_hm <- top$genes

#--> Get expression matrix
counts_mrna <- edgeR::cpm(tmm_set, normalized.lib.sizes = TRUE, log = TRUE)
sel_hm_mrna <- counts_mrna[match(sel_hm, row.names(counts_mrna), nomatch = 0),]

library(cluster)
library(factoextra)

#create plot of number of clusters vs total within sum of squares
#fviz_nbclust(counts_mrna, kmeans, method = "wss")

#--> Perform K-means clustering
clustering_hc <-  hclust(as.dist(1-cor(t(sel_hm_mrna))), method = "ward.D2")
clusters <- cutree(tree = clustering_hc , k = 2)

#--> Order genes
sel_hm_mrna <- sel_hm_mrna[clustering_hc$order,]

#--> Create row annotation matrix
annot_row = data.frame(row.names = row.names(sel_hm_mrna),
                       cluster = as.factor(clusters[clustering_hc$order]))

#--> Create column annotation matrix
pdata_chip <- pdata
colnames(sel_hm_mrna) <- row.names(pdata_chip)

#--> Build the heatmap
pheatmap(sel_hm_mrna,
         scale = "row",
         cluster_rows = FALSE,
         cluster_cols = TRUE, show_rownames = FALSE,
         clustering_method  = "ward.D2",
         clustering_distance_cols = "correlation",
         annotation_row = annot_row["cluster"],
         annotation_col = pdata[2],
         main = "Clustering ",
         color = viridis(100),
         labels_row = NA,
         fontsize = 6)
```
####### Table
```{r, message=FALSE}
library(DT)

table <- merge(top, annot_row, by = "row.names")

datatable(table[,-c(1,12,13)], extensions = "Buttons",
          options = list(buttons = c("csv", "excel"))) %>%
  formatRound(columns = c(3:9), digits=2)
```
####### Peak Annotation
```{r, message=FALSE}
library(ChIPpeakAnno)
library(EnsDb.Hsapiens.v86)
library(org.Hs.eg.db)
library(biomaRt)

gr_upreg <- GRanges(
    seqnames = Rle(up_loci_Grange$Chr),
    ranges = IRanges(start = up_loci_Grange$Start, end = up_loci_Grange$End),
    strand = Rle(strand(up_loci_Grange$Strand)))

gr_downreg <- GRanges(
    seqnames = Rle(down_loci_Grange$Chr),
    ranges = IRanges(start = down_loci_Grange$Start, end = down_loci_Grange$End),
    strand = Rle(strand(down_loci_Grange$Strand)))

ensembl_hs86_transcript <- transcripts(EnsDb.Hsapiens.v86)

macs_peak_ensembl_upreg <- annotatePeakInBatch(gr_upreg, 
                                         AnnotationData = ensembl_hs86_transcript)

macs_peak_ensembl_downreg <- annotatePeakInBatch(gr_downreg, 
                                         AnnotationData = ensembl_hs86_transcript)


mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL",
                dataset = "hsapiens_gene_ensembl")

macs_peak_ensembl_upregID <- addGeneIDs(annotatedPeak = macs_peak_ensembl_upreg, 
                                mart = mart,
                                feature_id_type = "ensembl_transcript_id",
                                IDs2Add = "hgnc_symbol")

macs_peak_ensembl_downregID <- addGeneIDs(annotatedPeak = macs_peak_ensembl_downreg, 
                                mart = mart,
                                feature_id_type = "ensembl_transcript_id",
                                IDs2Add = "hgnc_symbol")

genes_upreg <- macs_peak_ensembl_upregID$hgnc_symbol

genes_downreg <- macs_peak_ensembl_downregID$hgnc_symbol
```
## Comparison with differential analysis
```{r, message=FALSE}
genes_Stepniak_upreg_metabo <- read.csv("../RNA/upregulated_metabo_genes_DA.csv", header = TRUE)
genes_Stepniak_downreg_metabo <- read.csv("../RNA/downregulated_metabo_genes_DA.csv", header = TRUE)

genes_Jigna_downreg_metabo <- read.csv("../../Differential_analysis_Jigna/downregulated_metabo_genes_Jigna_samples.csv", header = TRUE)
genes_Jigna_upreg_metabo <- read.csv("../../Differential_analysis_Jigna/upregulated_metabo_genes_Jigna_samples.csv", header = TRUE)

list_genes_diff_analysis_1 <- append(genes_Jigna_downreg_metabo$x, genes_Jigna_upreg_metabo$x)
list_genes_diff_analysis_2 <- append(genes_Stepniak_downreg_metabo$x, genes_Stepniak_upreg_metabo$x)
unlist(list_genes_diff_analysis_1)
unlist(list_genes_diff_analysis_2)

list_genes_diff_analysis <- append(list_genes_diff_analysis_1,list_genes_diff_analysis_2)


loci_upreg <- intersect(list_genes_diff_analysis, genes_upreg)
loci_downreg <- intersect(list_genes_diff_analysis,genes_downreg)

metabo_genes_all <- read.table("../../Lists_genes/all_metabo.tsv", header = TRUE, sep = "\t")
metabo_genes <- metabo_genes_all$Gene_name

metabo_up <- intersect(metabo_genes,genes_upreg)
metabo_down <- intersect(metabo_genes,genes_downreg)

```
